File: merge.py
import requests
import re
import os
import time
import hashlib
from collections import Counter

# --- KONFIGURATION ---
SOURCES_FILE = "sources.txt"
OUTPUT_FILE = "blocklist.txt"
VERSION_FILE = "version.txt"
PROTECTED_KEYWORDS = ["oisd", "hagezi", "stevenblack", "firebog", "adaway", "badmojr"]
DOMAIN_REGEX = r"^(?:0\.0\.0\.0|127\.0\.0\.1)?\s*([a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+)"

SUBDOMAIN_THRESHOLD = 10 
MAX_FAILS = 3 

def is_subdomain(domain, domain_set):
    parts = domain.split('.')
    for i in range(len(parts) - 1, 1, -1):
        parent = ".".join(parts[len(parts)-i:])
        if parent in domain_set: return True
    return False

def get_parent_domain(domain):
    parts = domain.split('.')
    return ".".join(parts[-2:]) if len(parts) > 2 else domain

def main():
    start_time = time.time()
    print("--- OPTIMIZER START (Full Mode) ---")

    if not os.path.exists(SOURCES_FILE):
        print(f"ERROR: {SOURCES_FILE} nicht gefunden.")
        return

    # 1. Quellen einlesen
    with open(SOURCES_FILE, "r") as f:
        raw_lines = [l.strip() for l in f if l.strip() and not l.startswith("#")]
    
    url_map = {}
    for line in raw_lines:
        clean_url = re.sub(r"^(MASTER\||FAILx\d+\|)+", "", line).lower().strip("/")
        if clean_url not in url_map or "MASTER|" in line:
            url_map[clean_url] = line
    
    unique_urls = list(url_map.values())
    unique_urls.sort(key=lambda x: (not ("MASTER|" in x), x.lower()))

    all_domains = set()
    content_hashes = {}
    cleaned_sources = []
    total_raw_domains = 0 
    
    print(f"\n{'STATUS':<8} | {'NEUE':>10} | {'QUELLE'}")
    print("-" * 110)

    for line in unique_urls:
        is_master = "MASTER|" in line
        fail_match = re.search(r"FAILx(\d+)\|", line)
        current_fails = int(fail_match.group(1)) if fail_match else 0
        
        url_to_fetch = re.sub(r"^(MASTER\||FAILx\d+\|)+", "", line)
        
        # Kleine Pause um Server zu schonen
        time.sleep(0.5)

        try:
            r = requests.get(url_to_fetch, timeout=60, headers={'User-Agent': 'Mozilla/5.0'})
            if r.status_code == 200:
                m = hashlib.md5(r.text.encode('utf-8')).hexdigest()
                if m in content_hashes and not is_master:
                    print(f"{'MIRROR':<8} | {0:>10} | {url_to_fetch}")
                    continue

                current_list = set()
                for d_line in r.text.splitlines():
                    d_match = re.search(DOMAIN_REGEX, d_line)
                    if d_match:
                        current_list.add(d_match.group(1).lower())
                
                total_raw_domains += len(current_list)
                temp_new = [d for d in current_list if d not in all_domains and not is_subdomain(d, all_domains)]
                
                if len(temp_new) > 0 or is_master:
                    all_domains.update(temp_new)
                    content_hashes[m] = url_to_fetch
                    cleaned_sources.append(f"MASTER|{url_to_fetch}" if is_master else url_to_fetch)
                    status = "MASTER" if is_master else "OK"
                    print(f"{status:<8} | {len(temp_new):>10} | {url_to_fetch}")
                else:
                    print(f"{'REDUND.':<8} | {0:>10} | {url_to_fetch}")
            else: raise Exception()

        except Exception:
            new_fails = current_fails + 1
            is_protected = any(k in url_to_fetch.lower() for k in PROTECTED_KEYWORDS)
            if new_fails < MAX_FAILS or is_master or is_protected:
                prefix = "MASTER|" if is_master else ""
                cleaned_sources.append(f"{prefix}FAILx{new_fails}|{url_to_fetch}")
                print(f"{'OFFLINE':<8} | {'Try ' + str(new_fails):>10} | {url_to_fetch}")
            else:
                print(f"{'DEAD':<8} | {'REMOVED':>10} | {url_to_fetch}")

    # 2. TLD-AGGREGATION
    print("\nOptimiere Subdomains (Aggregation)...")
    parent_counts = Counter(get_parent_domain(d) for d in all_domains)
    auto_wildcards = {dom for dom, count in parent_counts.items() if count >= SUBDOMAIN_THRESHOLD}
    
    final_set = set()
    final_set.update(auto_wildcards)
    for d in all_domains:
        if get_parent_domain(d) not in auto_wildcards:
            if not is_subdomain(d, final_set):
                final_set.add(d)

    # 3. Speichern der Blockliste
    with open(OUTPUT_FILE, "w") as f:
        f.write(f"# Optimized Blocklist\n# Total Domains: {len(final_set)}\n")
        for d in sorted(final_set): f.write(d + "\n")

    # 4. Speichern der gesäuberten Quellen
    with open(SOURCES_FILE, "w") as f:
        f.write("# Cleaned Sources\n")
        for s in cleaned_sources: f.write(s + "\n")

    # --- DETAILLIERTE STATISTIK ---
    final_count = len(final_set)
    removed_count = total_raw_domains - final_count
    reduction_pct = (removed_count / total_raw_domains * 100) if total_raw_domains > 0 else 0
    duration = time.time() - start_time

    print("-" * 110)
    print(f"ZUSAMMENFASSUNG:")
    print(f"  - Brutto-Domains (Rohdaten):      {total_raw_domains:,}".replace(",", "."))
    print(f"  - Netto-Domains (blocklist.txt):  {final_count:,}".replace(",", "."))
    print(f"  - Davon Auto-Wildcards (TLDs):    {len(auto_wildcards):,}".replace(",", "."))
    print(f"  - Ersparnis (Müll & Aggregation): {removed_count:,}".replace(",", "."))
    print(f"  - Effizienz-Steigerung:           {reduction_pct:.2f}%")
    print(f"  - Bearbeitungszeit:               {duration:.2f} Sekunden")
    print("-" * 110)

    # --- NEU: VERSION.TXT FÜR GITHUB ACTIONS EXPORTIEREN ---
    try:
        with open(VERSION_FILE, "w") as f:
            f.write(f"Last Update: {time.strftime('%Y-%m-%d %H:%M')}\n")
            f.write(f"Total Domains: {final_count}\n")
        print(f"Statistik in {VERSION_FILE} gespeichert.")
    except Exception as e:
        print(f"Fehler beim Schreiben der version.txt: {e}")

    print("--- OPTIMIZER BEENDET ---")

if __name__ == "__main__":
    main()

File: merge.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/go-resty/resty/v2"
)

const domainRegex = `(?m)^(?:0\.0\.0\.0|127\.0\.0\.1)?\s*([a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+)`

var (
	threshold     = flag.Int("threshold", 50, "Subdomains ab denen per Wildcard geblockt wird")
	sourcesPath   = flag.String("sources", "sources.txt", "Datei mit Blocklisten-Quellen")
	outputPath    = flag.String("output", "blocklist.txt", "Ausgabedatei Blockliste")
	versionPath   = flag.String("version", "version.txt", "Ausgabedatei Versionsinfo")
	whitelistPath = flag.String("whitelist", "whitelist.txt", "Whitelist-Datei")
	allowlistPath = flag.String("allowlist", "allowlist.txt", "Ausgabedatei Allowliste")
	customPath    = flag.String("custom", "blocklist.custom.txt", "Optionale Datei mit eigenen Domains")
)

// cleanAndLoadWhitelist räumt die Whitelist auf und lädt sie in eine Map
func cleanAndLoadWhitelist(filename string) (map[string]bool, []string) {
	uniqueDomains := make(map[string]bool)
	var order []string

	file, err := os.Open(filename)
	if err != nil {
		return uniqueDomains, order
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		line = strings.Split(line, "#")[0] // Kommentare entfernen
		line = strings.TrimSpace(line)

		if line != "" {
			dom := strings.ToLower(line)
			if !uniqueDomains[dom] {
				uniqueDomains[dom] = true
				order = append(order, dom)
			}
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("Fehler beim Lesen von %s: %v\n", filename, err)
	}
	// Whitelist alphabetisch sortieren
	sort.Strings(order)
	// Die Datei direkt sauber wieder speichern
	if err := os.WriteFile(filename, []byte(strings.Join(order, "\n")+"\n"), 0644); err != nil {
		fmt.Printf("Warnung: Konnte %s nicht schreiben: %v\n", filename, err)
	}

	return uniqueDomains, order
}

func isIPAddress(s string) bool {
	parts := strings.Split(s, ".")
	if len(parts) != 4 {
		return false
	}
	for _, p := range parts {
		n, err := strconv.Atoi(p)
		if err != nil || n < 0 || n > 255 {
			return false
		}
	}
	return true
}

var multiPartTLDs = map[string]bool{
	"co.uk": true, "org.uk": true, "ac.uk": true, "gov.uk": true,
	"me.uk": true, "net.uk": true, "plc.uk": true, "sch.uk": true, "ltd.uk": true,
	"com.au": true, "net.au": true, "org.au": true, "edu.au": true, "gov.au": true,
	"co.jp": true, "ne.jp": true, "or.jp": true,
	"co.nz": true, "net.nz": true, "org.nz": true,
	"com.br": true, "org.br": true, "net.br": true,
	"co.kr": true, "or.kr": true, "ne.kr": true,
	"com.cn": true, "net.cn": true, "org.cn": true,
	"co.in": true, "net.in": true, "org.in": true,
	"com.mx": true, "org.mx": true, "net.mx": true,
	"co.za": true, "org.za": true, "net.za": true,
	"co.il": true, "org.il": true, "net.il": true,
	"com.pl": true, "net.pl": true, "org.pl": true,
	"com.pt": true, "net.pt": true, "org.pt": true,
	"com.ru": true, "net.ru": true, "org.ru": true,
	"com.sg": true, "edu.sg": true, "gov.sg": true, "net.sg": true, "org.sg": true,
	"com.tr": true, "net.tr": true, "org.tr": true, "gov.tr": true,
	"com.ar": true,
}

func isValidDomain(domain string) bool {
	if !strings.Contains(domain, ".") {
		return false
	}
	if isIPAddress(domain) {
		return false
	}
	if strings.HasSuffix(domain, ".local") || strings.HasSuffix(domain, ".lan") || strings.HasSuffix(domain, ".home.arpa") {
		return false
	}
	return true
}

func isWhitelisted(domain string, whitelist map[string]bool) bool {
	if whitelist[domain] {
		return true
	}
	parts := strings.Split(domain, ".")
	// Prüft hierarchisch (z.B. für cdn.example.com auch example.com)
	for i := 1; i < len(parts)-1; i++ {
		parent := strings.Join(parts[i:], ".")
		if whitelist[parent] {
			return true
		}
	}
	return false
}

func getParent(domain string) string {
	parts := strings.Split(domain, ".")
	if len(parts) < 2 {
		return domain
	}
	if len(parts) >= 3 {
		lastTwo := strings.Join(parts[len(parts)-2:], ".")
		if multiPartTLDs[lastTwo] {
			return strings.Join(parts[len(parts)-3:], ".")
		}
	}
	return strings.Join(parts[len(parts)-2:], ".")
}

func main() {
	flag.Parse()
	startTime := time.Now()
	fmt.Println("--- GO OPTIMIZER START (Parallel + Auto-Cleanup) ---")

	// 0. Whitelist laden & säubern
	whitelist, whitelistOrder := cleanAndLoadWhitelist(*whitelistPath)
	var whitelistHitCount int
	var hitMu sync.Mutex

	fmt.Printf("-> %d Domains in Whitelist (bereinigt).\n\n", len(whitelist))

	// 1. Quellen einlesen
	f, err := os.Open(*sourcesPath)
	if err != nil {
		fmt.Printf("Fehler: %s fehlt.\n", *sourcesPath)
		return
	}
	var uniqueSources []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" && !strings.HasPrefix(line, "#") {
			uniqueSources = append(uniqueSources, line)
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Printf("Fehler beim Lesen von %s: %v\n", *sourcesPath, err)
	}
	f.Close()

	allDomains := make(map[string]bool)
	var mu sync.Mutex
	var wg sync.WaitGroup
	re := regexp.MustCompile(domainRegex)
	prefixRe := regexp.MustCompile(`^(MASTER\||FAILx\d+\|)+`)

	client := resty.New().SetTimeout(60 * time.Second).SetRetryCount(3)

	fmt.Printf("%-8s | %-10s | %s\n", "STATUS", "NEUE", "QUELLE")
	fmt.Println(strings.Repeat("-", 80))

	for _, source := range uniqueSources {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			cleanURL := prefixRe.ReplaceAllString(url, "")

			resp, err := client.R().Get(cleanURL)
			if err != nil || resp.IsError() {
				fmt.Printf("%-8s | %-10s | %s\n", "OFFLINE", "Error", cleanURL)
				return
			}

			matches := re.FindAllStringSubmatch(resp.String(), -1)
			localNewCount := 0

			for _, m := range matches {
				dom := strings.ToLower(m[1])
				if !isValidDomain(dom) {
					continue
				}
				if isWhitelisted(dom, whitelist) {
					hitMu.Lock()
					whitelistHitCount++
					hitMu.Unlock()
					continue
				}
				mu.Lock()
				if !allDomains[dom] {
					allDomains[dom] = true
					localNewCount++
				}
				mu.Unlock()
			}
			fmt.Printf("%-8s | %-10d | %s\n", "OK", localNewCount, cleanURL)
		}(source)
	}
	wg.Wait()

	// 1b. Custom lokale Domains mergen
	if custom, err := os.ReadFile(*customPath); err == nil {
		added := 0
		for _, line := range strings.Split(string(custom), "\n") {
			line = strings.TrimSpace(line)
			line = strings.Split(line, "#")[0]
			line = strings.TrimSpace(line)
			if line != "" {
				dom := strings.ToLower(line)
				if isValidDomain(dom) && !allDomains[dom] {
					allDomains[dom] = true
					added++
				}
			}
		}
		if added > 0 {
			fmt.Printf("\n-> %d Domains aus %s übernommen.\n", added, *customPath)
		}
	}

	// 2. Aggregation (Subdomain-Konsolidierung)
	parentCounts := make(map[string]int)
	for dom := range allDomains {
		parentCounts[getParent(dom)]++
	}
	autoWildcards := make(map[string]bool)
	for p, count := range parentCounts {
		if count >= *threshold {
			autoWildcards[p] = true
		}
	}
	finalList := []string{}
	for dom := range allDomains {
		if !autoWildcards[getParent(dom)] {
			finalList = append(finalList, dom)
		}
	}
	for aw := range autoWildcards {
		finalList = append(finalList, aw)
	}
	sort.Strings(finalList)

	// --- DATEIEN SCHREIBEN ---
	out, err := os.Create(*outputPath)
	if err != nil {
		fmt.Printf("Fehler beim Erstellen von %s: %v\n", *outputPath, err)
		return
	}
	fmt.Fprintf(out, "# Optimized Blocklist\n# Total: %d\n", len(finalList))
	for _, d := range finalList {
		out.WriteString(d + "\n")
	}
	out.Close()

	allowOut, err := os.Create(*allowlistPath)
	if err != nil {
		fmt.Printf("Fehler beim Erstellen von %s: %v\n", *allowlistPath, err)
		return
	}
	fmt.Fprint(allowOut, "# Technitium Allow List\n")
	for _, dom := range whitelistOrder {
		allowOut.WriteString(dom + "\n")
	}
	allowOut.Close()

	timestamp := time.Now().Format("2006-01-02 15:04")
	finalCount := len(finalList)
	duration := time.Since(startTime)
	rawTotal := 0
	for _, c := range parentCounts {
		rawTotal += c
	}

	vFile, err := os.Create(*versionPath)
	if err != nil {
		fmt.Printf("Fehler beim Erstellen von %s: %v\n", *versionPath, err)
		return
	}
	fmt.Fprintf(vFile, "Last Update: %s\nTotal: %d\nWhitelist: %d\nEngine: Go (Resty)", timestamp, finalCount, whitelistHitCount)
	vFile.Close()

	jsonFile, err := os.Create("version.json")
	if err != nil {
		fmt.Printf("Fehler beim Erstellen von version.json: %v\n", err)
		return
	}
	fmt.Fprintf(jsonFile, `{"LastUpdate": "%s", "Total": %d, "Whitelist": %d}`, timestamp, finalCount, whitelistHitCount)
	jsonFile.Close()

	fmt.Println(strings.Repeat("-", 80))
	fmt.Printf("ZUSAMMENFASSUNG:\n")
	fmt.Printf("Zeit: %v\n", duration)
	fmt.Printf("Whitelist: %d Einträge, %d Hits\n", len(whitelist), whitelistHitCount)
	fmt.Printf("Quellen: %d\n", len(uniqueSources))
	fmt.Printf("Domains extrahiert: %d | Nach Dedup: %d | Wildcards: %d | Final: %d\n", rawTotal, len(allDomains), len(autoWildcards), finalCount)
}
